store/src/java/com/zimbra/cs/store/StoreManager.java

```diff
@@ -68,18 +92,101 @@ public static StoreManager getInstance(String className) {
+        }
+        throw ServiceException.OPERATION_DENIED("Not support when Running Imap Inside Mailboxd");
+    }
 
     /**
      * Used for unit testing.
      */
     public static void setInstance(StoreManager instance) {
         synchronized (StoreManager.class) {
-            ZimbraLog.store.info("Setting StoreManager to " + instance.getClass().getName());
+            if (instance != null) {
+                ZimbraLog.store.info("Setting StoreManager to " + instance.getClass().getName());
+                sInstance.shutdown();
+            }
+            ZimbraLog.store.info("Setting StoreManager to " + instance);
             sInstance = instance;
         }
     }
 
+    /**
+     * Get reader StoreManager for the locator
+     * TODO: add documentation here
+     * @param locator
+     * @return
+     */
+    public static StoreManager getReaderSMInstance(String locator) {
+        try {
+            // TODO: check if calling Provision is feasible for realtime calls, otherwise LC should be used here
+            boolean multiReaderSMEnabled = Provisioning.getInstance().getLocalServer().isMultiReaderSMEnabled();
+            if (multiReaderSMEnabled && !StringUtil.isNullOrEmpty(locator)) {
+                Optional<Short> volumeId = MailItemHelper.findMyVolumeId(locator);
+                if (volumeId.isPresent()) {
+                    ZimbraLog.store.error("Volume for locator: %s volumeId: %s", locator, volumeId.get());
+                    StoreManager storeManager = StoreManager.getStoreManagerForVolume(volumeId.get(), true);
+                    if (null != storeManager) {
+                        ZimbraLog.store.error("Store Manager for %s Volume: %s ", storeManager.getClass().getSimpleName(), volumeId.get());
+                        return storeManager;
+                    }
+                    ZimbraLog.store.debug("not able to load reader store manager for volumeId: %s", volumeId);
+                }
+            }
+        } catch (ServiceException e) {
+            ZimbraLog.store.error("Error while loading StoreManager for locator %s", locator, e);
+        }
+        ZimbraLog.store.info("Fallback: master StoreManager will be used for reading");
+        return getInstance();
+    }
+
+    public static StoreManager getStoreManagerForVolume(Short volumeId, boolean useCached) {
+        if (useCached && volumeStoreManagerCache.containsKey(volumeId)) {
+            ZimbraLog.store.error("Store Manager cached for %s", volumeId);
+            return volumeStoreManagerCache.get(volumeId);
+        }
+        StoreManager storeManager = null;
+        try {
+            Volume volume = VolumeManager.getInstance().getVolume(volumeId);
+            String className = volume.getStoreManagerClass();
+            ZimbraLog.store.error("loading Store Manager: %s  for %s", className, volumeId);
+            storeManager = (StoreManager) ClassHelper.getZimbraClassInstanceBy(className);
+            if (null != storeManager) {
```

Comment:
null check is better as `storeManager != null`

Reply:
I feel that more faster but its less readable